collectors/modality-probe-collector-common/src/lib.rs

```diff
@@ -283,320 +354,414 @@ impl TryFrom<&LogFileRow> for LogEntry {
+                        );
+                    } else if raw_entry.has_event_with_payload_bit_set() {
+                        interpret_next_as = Next::Payload(
+                            raw_entry
+                                .interpret_as_event_id()
+                                .ok_or_else(|| SerializationError)?,
+                        );
+                    } else {
+                        owned_report.event_log.push(EventLogEntry::Event(
+                            raw_entry
+                                .interpret_as_event_id()
+                                .ok_or_else(|| SerializationError)?,
+                        ));
+                    }
                 }
-                LogEvent::EventWithPayload(ev, payload) => {
-                    log_entries_buffer.push(LogEntry {
-                        session_id,
-                        segment_id: next_segment_id.into(),
-                        segment_index,
-                        probe_id,
-                        data: LogEntryData::EventWithPayload(*ev, *payload),
-                        receive_time,
-                    });
+                Next::Clock(id) => {
+                    let (epoch, ticks) = modality_probe::unpack_clock_word(raw);
+                    owned_report
+                        .event_log
+                        .push(EventLogEntry::TraceClock(LogicalClock { id, epoch, ticks }));
+                    interpret_next_as = Next::DontKnow;
+                }
+                Next::Payload(id) => {
+                    owned_report
+                        .event_log
+                        .push(EventLogEntry::EventWithPayload(id, raw));
+                    interpret_next_as = Next::DontKnow;
                 }
             }
-
-            segment_index += 1;
         }
-
-        next_segment_id += 1;
+        Ok(owned_report)
     }
-
-    next_segment_id
 }
 
-pub fn add_owned_report_to_entries(
-    report: LogReport,
-    session_id: SessionId,
-    initial_segment_id: SegmentId,
-    receive_time: DateTime<Utc>,
-    log_entries_buffer: &mut Vec<LogEntry>,
-) -> u32 {
-    let mut next_segment_id = initial_segment_id.0;
-    let probe_id = report.probe_id;
-    for segment in report.segments {
-        let mut segment_index = 0;
-
-        for clock_bucket in segment.clocks {
-            log_entries_buffer.push(LogEntry {
-                session_id,
-                segment_id: next_segment_id.into(),
-                segment_index,
-                probe_id,
-                data: LogEntryData::LogicalClock(
-                    clock_bucket.id,
-                    clock_bucket.epoch,
-                    clock_bucket.ticks,
-                ),
-                receive_time,
-            });
+enum Next {
+    Clock(ProbeId),
+    Payload(EventId),
+    DontKnow,
+}
 
-            segment_index += 1;
+impl Report {
+    pub fn write_into_le_bytes(&self, bytes: &mut [u8]) -> Result<usize, ReportWireError> {
```

Comment:
Is this expected to be used outside of tests?

Reply:
I would expect this to be useful if at some point someone wants to use this `Report` structure to synthesize a report outside of the context of the provided probe implementation.  For now, `#[cfg(test)]` is probably appropriate, though.